akka-actor/src/main/java/akka/dispatch/AbstractBoundedNodeQueue.java

```diff
@@ -167,9 +167,17 @@ public final T poll() {
             final Node<T> deq = getDeq();
             final Node<T> next = deq.next();
             if (next != null) {
+                //Can't rely on next.value after casDeq because 'next' can be popped by another thread already
+                //before this one finishes. 
+                //Thus reading and storing the value before casDeq is needed. 
+                T nextVal = next.value;
                 if (casDeq(deq, next)) {
-                    deq.value = next.value;
-                    next.value = null;
+                    deq.value = nextVal;
+                    
+                    //Only set this to null if next still contains the expected
+                    //value. If not, "deq.value = nextVal" has already happened on a different thread.
+                    //and we don't want to override that since that's another value popping.
+                    next.casNullValue(nextVal);
```

Comment:
You're right, didn't think of that. Though you're probably better off with the CAS compared to bluntly setting `next.value` to `null` since that possibly always incorrectly overwrites values.

This simply means that `next.value` can't be set to `null` in multiple consumer scenarios (without introducing more concurrency controls). The obvious other option is to accept a slight (temporary?) memory leak on the last node. Which actually doesn't sound that bad for a queue that is expected to move work around during it's lifetime.